core/src/test/java/com/sequenceiq/cloudbreak/service/upgrade/ccm/StackCcmUpgradeServiceTest.java

```diff
@@ -38,24 +53,115 @@ class StackCcmUpgradeServiceTest {
+
     @InjectMocks
     private StackCcmUpgradeService underTest;
 
+    @BeforeEach
+    void setUp() {
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = Tunnel.class, names = { "CCM", "CCMV2" }, mode = Mode.INCLUDE)
+    void testUpgradeCcm(Tunnel tunnel) {
+        Stack stack = createStack(tunnel, Status.AVAILABLE);
+        DetailedEnvironmentResponse environment = createEnvironment(Tunnel.latestUpgradeTarget());
+        when(stackService.getByNameOrCrnInWorkspace(eq(NameOrCrn.ofCrn("crn")), any())).thenReturn(stack);
+        when(environmentService.getByCrn("envCrn")).thenReturn(environment);
+
+        StackCcmUpgradeV4Response response = underTest.upgradeCcm(NameOrCrn.ofCrn("crn"));
+
+        ArgumentCaptor<UpgradeCcmTriggerRequest> requestCaptor = ArgumentCaptor.forClass(UpgradeCcmTriggerRequest.class);
+        verify(reactorNotifier).notify(eq(STACK_ID), eq("UPGRADE_CCM_TRIGGER_EVENT"), requestCaptor.capture());
+        UpgradeCcmTriggerRequest request = requestCaptor.getValue();
+        assertThat(request.getResourceId()).isEqualTo(STACK_ID);
+        assertThat(request.getClusterId()).isEqualTo(CLUSTER_ID);
+        assertThat(request.getOldTunnel()).isEqualTo(tunnel);
+        assertThat(response.getResponseType()).isEqualTo(CcmUpgradeResponseType.TRIGGERED);
+        assertThat(response.getResourceCrn()).isEqualTo("crn");
+    }
+
+    @Test
+    void environmentNotLatest() {
+        Stack stack = createStack(Tunnel.CCM, Status.AVAILABLE);
+        DetailedEnvironmentResponse environment = createEnvironment(Tunnel.DIRECT);
+        when(stackService.getByNameOrCrnInWorkspace(eq(NameOrCrn.ofCrn("crn")), any())).thenReturn(stack);
+        when(environmentService.getByCrn("envCrn")).thenReturn(environment);
+
+        assertThatThrownBy(() -> underTest.upgradeCcm(NameOrCrn.ofCrn("crn")))
+                .isInstanceOf(BadRequestException.class);
+
+        verifyNoInteractions(reactorNotifier);
+    }
+
     @Test
-    void testUpgradeCcm() {
+    void testAlreadyUpgraded() {
+        Stack stack = createStack(Tunnel.latestUpgradeTarget(), Status.AVAILABLE);
+        DetailedEnvironmentResponse environment = createEnvironment(Tunnel.latestUpgradeTarget());
+        when(stackService.getByNameOrCrnInWorkspace(eq(NameOrCrn.ofCrn("crn")), any())).thenReturn(stack);
+        when(environmentService.getByCrn("envCrn")).thenReturn(environment);
+
+        StackCcmUpgradeV4Response response = underTest.upgradeCcm(NameOrCrn.ofCrn("crn"));
+
+        verifyNoInteractions(reactorNotifier);
+        assertThat(response.getResponseType()).isEqualTo(CcmUpgradeResponseType.SKIP);
+        assertThat(response.getResourceCrn()).isEqualTo("crn");
+        assertThat(response.getFlowIdentifier()).isEqualTo(FlowIdentifier.notTriggered());
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = Status.class, names = { "AVAILABLE", "MAINTENANCE_MODE_ENABLED", "UPGRADE_CCM_FAILED" }, mode = Mode.EXCLUDE)
```

Comment:
Because stack.getStatus.isAvailable compares the current value against a set of Available values where the MAINTENANCE_MODE is among the set elements.

Reply:
Oh, I see.